### PR TITLE
Graph refactor and random-walks in chaoda

### PR DESCRIPTION
**Pull request by nishaq503**
YAY! This was very fun to code up

### CODE REVIEW

File: pyclam/manifold.py [hunk 1]
```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
...
-        """ An iterator over the clusters in the graph. """
-        yield from self.edges
+        yield from self.clusters
 
     def __str__(self) -> str:
-        # Cashing value because sort can be expensive on many clusters.
         if 'str' not in self.cache:
             self.cache['str'] = ', '.join(sorted(list(map(str, self.clusters))))
+        else:
+            pass
```
Comment (thoward27) [inline pyclam/manifold.py:519]: why else pass? Seems overly verbose

File: pyclam/manifold.py [hunk 2]
```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
...
         return self.cache['str']
 
     def __repr__(self) -> str:
-        # Cashing value because sort can be expensive on many clusters.
         if 'repr' not in self.cache:
             self.cache['repr'] = '\n'.join(sorted(list(map(repr, self.clusters))))
+        else:
+            pass
```
Comment (thoward27) [inline pyclam/manifold.py:526]: Not a fan of else pass pattern

File: pyclam/manifold.py [hunk 3]
```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
...
 
+    @property
+    def edges_dict(self) -> Dict[Cluster, Set[Edge]]:
+        if 'edges_dict' not in self.cache:
+            self.cache['edges_dict'] = {
+                cluster: {edge for edge in self.edges if cluster in edge}
+                for cluster in self.clusters
+            }
+        else:
+            pass
```
Comment (thoward27) [inline pyclam/manifold.py:543]: Like this less every time I see it

File: pyclam/manifold.py [hunk 4]
```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
...
     def metric(self) -> Metric:
-        return next(iter(self.edges)).metric
+        return next(iter(self.clusters)).metric
 
     @property
     def depth(self) -> int:
         if 'depth' not in self.cache:
             self.cache['depth'] = max((cluster.depth for cluster in self.clusters))
+        else:
+            pass
```
Comment (thoward27) [inline pyclam/manifold.py:567]: 💣

File: pyclam/manifold.py [hunk 5]
```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
...
         return self.cache['depth']
 
     @property
-    def clusters(self) -> Iterable[Cluster]:
-        return self.edges.keys()
+    def min_depth(self) -> int:
+        if 'min_depth' not in self.cache:
+            self.cache['min_depth'] = min((cluster.depth for cluster in self.clusters))
+        else:
+            pass
```
Comment (thoward27) [inline pyclam/manifold.py:575]: 😨

File: pyclam/manifold.py [hunk 6]
```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
...
-    def walkable_clusters(self) -> Set[Cluster]:
-        """ Set of Clusters not subsumed by other clusters. """
-        if 'walkable_clusters' not in self.cache:
+    def subgraphs(self) -> Set['Graph']:
+        """ Returns the set of all connected component subgraphs in the Graph. """
+        if not self.is_built:
             self.build_edges()
-        return self.cache['walkable_clusters']
+        else:
+            pass
```
Comment (thoward27) [inline pyclam/manifold.py:588]: 👻

File: pyclam/manifold.py [hunk 7]
```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
...
 
     @property
-    def walkable_clusters(self) -> Set[Cluster]:
-        """ Set of Clusters not subsumed by other clusters. """
-        if 'walkable_clusters' not in self.cache:
+    def subgraphs(self) -> Set['Graph']:
+        """ Returns the set of all connected component subgraphs in the Graph. """
+        if not self.is_built:
             self.build_edges()
-        return self.cache['walkable_clusters']
```
Comment (thoward27) [inline pyclam/manifold.py:527]: I don't see how this function returns anything

**Review by thoward27 — CHANGES_REQUESTED** [review]
I only got part of the way through this, but I am strongly against this else pass pattern thing

File: pyclam/manifold.py (hunk 1 above)
Comment (nishaq503) [reply pyclam/manifold.py:519]: It's a Rust habit I subconsciously brought over. I didn't even notice it. Exhaustive if-branches are required in Rust and can be bad form in Python. I'm taking them all out of here

File: pyclam/manifold.py (hunk 7 above)
Comment (nishaq503) [reply pyclam/manifold.py:527]: Does line 589 in the newest push clear things up? I'm only guessing at what meant here...

**Comment by thoward27** [comment]
@nishaq503 can you add a little context here, what is the goal of this PR? I see a lot of code got axed, does that functionality now live elsewhere? Or, are we just dropping those bits for now?

File: pyclam/manifold.py [hunk 8]
```diff
@@ -517,600 +568,236 @@ def depth(self) -> int:
...
-        # visit all reachable walkable clusters
+
         while frontier:
             visited.update(frontier)
-            frontier = {
-                neighbor for cluster in frontier
-                for neighbor in self.neighbors(cluster, choice='walkable')
-                if neighbor not in visited
-            }
+            frontier = {neighbor for cluster in frontier for neighbor in self.neighbors(cluster) if neighbor not in visited}
```
Comment (thoward27) [inline pyclam/manifold.py:712]: !autoformat me

File: pyclam/manifold.py [hunk 9]
```diff
@@ -517,600 +568,236 @@ def depth(self) -> int:
...
 
+    def _add(self, cluster: Cluster) -> None:
+        """ Adds the cluster and all associated edges to the graph.
+
+        Assumes that the cluster is not already in the graph.
+        Caller need to invalidate cache.
+        """
+        self._find_neighbors(cluster)
+        self_edges = {edge for edge in self.edges if edge.to_self}
+        self.edges -= self_edges
```
Comment (thoward27) [inline pyclam/manifold.py:762]: don't forget to delete this once you fix find neighbors

File: pyclam/manifold.py [hunk 10]
```diff
@@ -517,600 +568,236 @@ def depth(self) -> int:
...
+        return
+
+    def _remove(self, cluster: Cluster) -> None:
+        """ Removes the cluster and all associated edges from the graph.
+
+        Assumes that the cluster is in the graph.
+        Caller need to invalidate cache.
+        """
+        removed_edges = {edge for edge in self.edges if cluster in edge}
+        self.edges -= removed_edges
```
Comment (thoward27) [inline pyclam/manifold.py:772]: use edges_from

File: pyclam/tests/test_functional.py [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 import numpy as np
 
-from pyclam import Manifold, criterion
+from pyclam import criterion, Manifold
```
Comment (thoward27) [inline pyclam/tests/test_functional.py:5]: just beautiful

File: pyclam/types.py [hunk 12]
```diff
@@ -8,5 +8,3 @@
 Vector = List[int]
 DistanceFunc = Callable[[Data, Data], Radius]
 Metric = Union[str, DistanceFunc]
-Edge = namedtuple('Edge', 'neighbor distance probability')
-CacheEdge = namedtuple('CacheEdge', 'source neighbor distance probability')
```
Comment (thoward27) [inline pyclam/types.py:12]: 😢 so sad to see us not use this garbage

**Review by thoward27 — APPROVED** [review]
LGTM with some comments

**Pull request merged**